[ALLUXIO-1886] Add Kafka Connect to export stream data to alluxio

zhutongxue opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-1886
Add kafka connector code in Intergration module

AmplabJenkins commented:
Can one of the admins verify this patch?

gpang commented:
ok to test

gpang commented:
Jenkins, add to whitelist

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9350/<h2>Build result: FAILURE</span></h2>[...truncated 2972 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/target/alluxio-keyvalue-client-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-client/1.1.0-SNAPSHOT/alluxio-keyvalue-client-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/target/alluxio-core-server-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/target/alluxio-core-server-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/target/alluxio-core-server-1.1.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.1.0-SNAPSHOT/alluxio-underfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.1.0-SNAPSHOT/alluxio-integration-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.1.0-SNAPSHOT/alluxio-keyvalue-1.1.0-SNAPSHOT.pomchannel stopped
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9361/
Test PASSed.

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnector.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * This class is a Kafka Connect Connector implementation that exports data from Kafka to Alluxio.
+ */
+public class AlluxioSinkConnector extends Connector {
```
NIT: @zhutongxue in Alluxio codebase, we typically mark classes `final` if it is not intended to have subclasses.

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnector.java:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  @Override public void stop() {
+
```
NIT: unnecessary empty line.

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnector.java:
```diff
@@ -0,0 +1,54 @@
+
+  @Override public List<Map<String, String>> taskConfigs(int maxTasks) {
+    List<Map<String, String>> taskConfigs = new ArrayList<Map<String, String>>();
+    Map<String, String> taskProps = new HashMap<String, String>();
```
NIT: a convention in alluxio codebase: you can have
```java
Map<String, String> taskProps = new HashMap<>();
```
since we now moved to Java 7

apc999 commented on integration/kafka/pom.xml:
```diff
@@ -0,0 +1,81 @@
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
```
NIT: for `xml` files, Alluxio convention is to use 2-space convention. 
see `pom.xml` under the root dir.

apc999 commented on integration/kafka/pom.xml:
```diff
@@ -0,0 +1,81 @@
+    </repositories>
+
+    <dependencies>
+        <dependency>
```
NIT: better to keep the dependencies sorted alphabetically, e.g., sort by `groupId` first, then `artifactId`.

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTaskConstants.java:
```diff
@@ -0,0 +1,22 @@
+ */
+public class AlluxioSinkTaskConstants {
+
+  public static final int OFFSET_LENGTH = 10;
```
I feel these two public constants can be defined inside `AlluxioSinkTask`.
this is not critical, up to you

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioTopicPartitionWriter.java:
```diff
@@ -0,0 +1,242 @@
+public class AlluxioTopicPartitionWriter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioSinkTask.class);
+
```
in alluxio code base, we usually keep the source compact. I think we could remove the empty line between every two definition of member variables.

apc999 replied:
also for those member variables, if they are only assigned in constructor and never changed, we typically mark them `final`; otherwise if they are possibly changed in other methods, leave them as is.

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/format/AlluxioFormat.java:
```diff
@@ -0,0 +1,40 @@
+public interface AlluxioFormat {
+
+  /**
+   * Get File extension name.
```
NIT: Get -> Gets, and also below, Write -> Writes.
As an Alluxio code convention, we tend to use present tense verbs in Javadoc

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/format/AlluxioFormat.java:
```diff
@@ -0,0 +1,40 @@
+   *
+   * @return File extension name
+   */
+  public String getExtension();
```
NIT: inside an interface definition, we typically don't have `public` as they are redundant modifiers on interfaces

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnectorConfig.java:
```diff
@@ -0,0 +1,67 @@
+          ROTATION_RECORD_NUM_DOC)
+      .define(ROTATION_TIME_INTERVAL, Type.LONG, "-1", ConfigDef.Importance.LOW,
+          ROTATION_TIME_INTERVAL_DOC)
+      .define(ALLUXIO_FORMAT, Type.STRING, "alluxio.connect.format.JsonFormat",
```
we can replace the constant string `"alluxio.connect.format.JsonFormat"` by `JsonFormat.class.getName()` for portability, rather than hardcoding the class name here.

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+      tpWriter.initialize();
+      mTpWriters.put(tp, tpWriter);
+    }
+
```
unnecessary empty line

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+   * @param offsets mapping of TopicPartition to committed offset
+   */
+  @Override public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) {
+
```
unnecessary empty line

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+   * Perform any cleanup to stop this task.
+   */
+  @Override public void stop() {
+
```
unnecessary empty line.

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+      mFormat = (AlluxioFormat) Class
+          .forName(mConfig.getString(AlluxioSinkConnectorConfig.ALLUXIO_FORMAT)).newInstance();
+    } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
+      LOG.error("Exception is {}", e);
```
Question: should we raise the exception here, if the `mFormat` can not be initialized ?
currently, we simply `LOG` it and proceed, is this fine?

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+  private AlluxioFormat mFormat;
+
+  private Map<TopicPartition, AlluxioTopicPartitionWriter> mTpWriters =
+      new HashMap<TopicPartition, AlluxioTopicPartitionWriter>();
```
NIT: `new HashMap<>();`

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+
+  }
+
+  @Override public void put(Collection<SinkRecord> records) {
```
NIT:
`@Override` should be on a separate line

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+      int partition = record.kafkaPartition();
+      TopicPartition topicPartition = new TopicPartition(topic, partition);
+      AlluxioTopicPartitionWriter writer = mTpWriters.get(topicPartition);
+      if (writer != null) {
```
if `writer == null`, should we log error at least?

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+   * @param partitions The list of partitions that are now assigned to the task (may include
+   *                   partitions previously assigned to the task)
+   */
+  public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
```
is this an override method?

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+   *
+   * @param partitions The list of partitions that should be closed
+   */
+  public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
```
same here, annotate `@Override` if it is

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,170 @@
+  /**
+   * Perform any cleanup to stop this task.
+   */
+  @Override public void stop() {
```
Alluxio convention: put annotation `@Override` on a separate line

apc999 commented:
@zhutongxue Thanks for the awesome PR. I made one pass with some questions and mostly some formatting/style issues. Please address them and update me before I make another pass.

Meanwhile, how possible to have unit tests for those classes?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9477/
Test PASSed.

zhutongxue commented:
@apc999 Thank you very much for your comments, the comments are very helpful to me. According to your advice, I have made some modification about formatting/style issues. Meanwhile I have changed some classes/variables type and some exception handling .

peisun1115 commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,111 @@
+    alluxio.format=alluxio.kafka.connect.format.JsonFormat
+
+
+User can configure connector class in this property file,the `topics` specify the topics we want export data from kafka. `alluxio.url` is the alluxio address where we export data to. `topics.dir ` is the top level directory in alluxio filesystem where the kafka data is stored.`rotation.num` specify the kafka record number writen in alluxio file before file closes.`rotation.time.interval` is simil ...
```
space after "file,"

peisun1115 commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,111 @@
+
+Confluent Connector and Worker configuration details can be seen [Kafka connector configuration](http://docs.confluent.io/2.0.1/connect/userguide.html#configuring-connectors)
+##Setup
+- Install and run Alluxio
```
Can you add a link here (e.g. http://alluxio.org/documentation/master/en/Getting-Started.html)

peisun1115 commented:
@zhutongxue  Thanks for making this integration!

I was only reading the ReadMe.md file and try to use it.  
These are what I did:
1. Quickstart on kafka connect (
http://docs.confluent.io/2.0.1/quickstart.html and http://docs.confluent.io/2.0.1/connect/intro.html)
2. Started Alluxio.
3.  At this point, I have kafla instance running (including schema registration worker, zookeeper, kafka server). I ran 
./bin/connect-standalone /Users/peis/workspace/tachyon/integration/kafka/alluxio-worker-standalone-example.properties  /Users/peis/workspace/tachyon/integration/kafka/alluxio-connect-example.properties

It failed because 

ERROR Stopping after connector error (org.apache.kafka.connect.cli.ConnectStandalone:91)
org.apache.kafka.common.config.ConfigException: Invalid value alluxio.kafka.connect.AlluxioSinkConnector for configuration connector.class: Class alluxio.kafka.connect.AlluxioSinkConnector could not be found.

I am sure I have done something wrong or missed some step :)  Could you point it out and improve your readme doc? 

Thank you

zhutongxue commented:
@peisun1115  Thank you for your comments.

According to your description, the error is caused by "alluxio.kafka.connect.AlluxioSinkConnector could not be found". The reason is that the classpath of connect-standalone command does not include AlluxioSinkConnector class, so the AlluxioSinkConnector class can't be found. 

I updated the Readme.md and added **How to Configure Alluxio connector run environment**, maybe it helps to you.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9483/
Test PASSed.

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+##Kafka Alluxio Connect
```
How about Kafka Alluxio Connector?

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+##Kafka Alluxio Connect
+Kafka Alluxio Connector implements exporting kafka's specific topic data to alluxio filesystem.User can define data format stored in alluxio file,such as JsonFormat.
```
Kafka Alluxio Connector exports data in Kafka to Alluxio. The connector supports user defined data format and shipped with JSON format.

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+##Kafka Alluxio Connect
+Kafka Alluxio Connector implements exporting kafka's specific topic data to alluxio filesystem.User can define data format stored in alluxio file,such as JsonFormat.
+##Perquisites
+- confluent-2.0.1
```
Confluent Platform 2.0.1 or above, may be provides a download link?

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+####Example
+Here is the content of `etc/kafka-connect-alluxio/alluxio-connect.properties`
+    
+	name=alluxio-sink
```
nit: alignment.

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+User can configure connector class in this property file, the `topics` specify the topics we want export data from kafka. `alluxio.url` is the alluxio address where we export data to. `topics.dir ` is the top level directory in alluxio filesystem where the kafka data is stored. `rotation.num` specify the kafka record number writen in alluxio file before file closes. `rotation.time.interval` is si ...
+###Connector Worker Configuration
+####Example
+Here is the sample configuration for a standalone Kafka Connect worker that uses Avro serialization and integrates the the Schema Registry. This sample configuration assumes a local installation of Confluent Platform with all services running on their default ports.
```
for running Kafka Connect in standalone mode

Maybe link to the documentation of Confluent Platform quick start on how to run the services? http://docs.confluent.io/2.0.1/quickstart.html

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+
+    	./bin/zookeeper-server-start etc/kafka/zookeeper.properties &
+    	./bin/schema-registry-start etc/schema-registry/schema-registry.properties
+    	./bin/kafka-server-start etc/kafka/server.properties
```
Kafka should be started before Schema Registry

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+	Note:schema-registry-start is opitional. If you need schema registry such as Avro schema registry, this command is essential.
+- Configure Alluxio connector run environment
+  
+	Create `kafka-connect-alluxio` dir in confluent `share/java` dir, put `alluxio-integration-kafka-connect` jar into newly-created `kafka-connect-alluxio` dir, then create `kafka-connect-alluxio` dir in confluent `etc` dir, put alluxio connector property files(such as `alluxio-connect.properties` etc.) into the dir.
```
Actually, you can just add the Alluxio connector to the CLASSPATH of Confluent Platform's bin scripts.  

export CLASSPATH=path to the Alluxio connector.

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+  
+	`./bin/connect-standalone etc/kafka-connect-alluxio/alluxio-worker-standalone.properties etc/kafka-connect-alluxio/alluxio-connect.properties `
+	
+- Test with avro console
```
Test with the Avro Console Producer.

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+	
+	If you use Avro format as producer input format, the fllowing configuration in worker properties is necessary, and you also make sure that schema-registry service has been started. 
+	
+	    key.converter=io.confluent.connect.avro.AvroConverter
```
nit: Alignment

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+    	{"f1": "value2"}
+    	{"f1": "value3"}
+
+	You should see a file with name `/topics/test_alluxio/partition=0/test_alluxio+0+0000000000+0000000002.json`. The file name is encoded as `topic+kafkaPartition+startOffset+endOoffset.format`, which refers to Kafka-Hdfs-Connector.
```
No need to mention the HDFS connector here.

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+
+	You should see a file with name `/topics/test_alluxio/partition=0/test_alluxio+0+0000000000+0000000002.json`. The file name is encoded as `topic+kafkaPartition+startOffset+endOoffset.format`, which refers to Kafka-Hdfs-Connector.
+
+- Test with standard producer console and Json format   
```
Maybe put this before the Avro part?

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+
+	You should see a file in alluxio with name `/topics/test_alluxio/partition=0/test_alluxio+0+0000000000+0000000002.json` as before.
+##Distributed Deployment
+To start distributed worker, create a worker configuration as standalone mode, and then execute ` connect-distributed ` command.
```
To run Kafka Connect in distributed mode, modify the con

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+    ./bin/connect-distributed etc/kafka-connect-alluxio/alluxio-worker-distributed.properties
+
+In distributed mode, You can use the REST API to manage the connectors running in the cluster. You can read [Connectors Configuration](http://http://docs.confluent.io/2.0.1/connect/userguide.html#getting-started "") for more details. Here is an example.
+
```
Would be nice to incluce the REST API reference
http://docs.confluent.io/2.0.1/connect/userguide.html#rest-interface

Ishiihara commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,107 @@
+
+	`curl -X POST -H "Content-Type: application/json" --data '{"name": "alluxio-sink", "config": {"connector.class":"alluxio.kafka.connect.AlluxioSinkConnector", "tasks.max":"1", "topics":"test_alluxio","alluxio.url":"alluxio://localhost:19998","topics.dir":"topics","rotation.num":"3","alluxio.format":"alluxio.kafka.connect.format.JsonFormat" }}' http://localhost:8083/connectors`
+
+ After post a connector to workers, you can send data to specific topic as standalone mode.
```
After the connector is created, the data in Kafka for the specified topics will be continuously export to Alluxio.

Ishiihara commented on integration/kafka/alluxio-worker-standalone-example.properties:
```diff
@@ -0,0 +1,42 @@
+value.converter=io.confluent.connect.avro.AvroConverter
+value.converter.schema.registry.url=http://localhost:8081
+
+#key.converter=org.apache.kafka.connect.json.JsonConverter
```
If we ends up putting the JSON converters in the comments, we should at least add some description on why we need it.

Ishiihara commented on integration/kafka/pom.xml:
```diff
@@ -0,0 +1,81 @@
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
I think we can do a better job of packing here. Essentially we don't need a fat jar, right? If possible, would you mind to follow how do we package in the HDFS connector?

Ishiihara commented on integration/kafka/pom.xml:
```diff
@@ -0,0 +1,81 @@
+    <dependency>
+      <groupId>io.confluent</groupId>
+      <artifactId>kafka-connect-avro-converter</artifactId>
+      <version>${confluent.version}</version>
```
We add some features after 2.0.1 release and 3.0.0 is coming out soon. May be you can develop against 3.0.0-SNAPSHOT?

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnector.java:
```diff
@@ -0,0 +1,58 @@
+
+  @Override
+  public String version() {
+    return AlluxioSinkConnectorConfig.ALLUXIO_CONNECTOR_VERSION;
```
In the HDFS and JDBC's connector implementation, we used the project version as the connector version. For example, if in the pom.xml, the project version 3.0.0, then the connector version will be 3.0.0. Using project version instead of providing version in the connector configuration, we have less chance to provide the wrong value.

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnectorConfig.java:
```diff
@@ -0,0 +1,61 @@
+  public static final String ALLUXIO_URL = "alluxio.url";
+  public static final String TOPICS_DIR = "topics.dir";
+  public static final String ROTATION_RECORD_NUM = "rotation.num";
+  public static final String ROTATION_TIME_INTERVAL = "rotation.time.interval";
```
Maybe using `rotation.time.interval.ms` to indicate time unit.

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnectorConfig.java:
```diff
@@ -0,0 +1,61 @@
+          ROTATION_RECORD_NUM_DOC)
+      .define(ROTATION_TIME_INTERVAL, Type.LONG, "-1", ConfigDef.Importance.LOW,
+          ROTATION_TIME_INTERVAL_DOC)
+      .define(ALLUXIO_FORMAT, Type.STRING, JsonFormat.class.getName(),
```
Also, it seems to pay off to define a static variable for each configurations default value, at least we explicitly express the type information.

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,177 @@
+      mFormat = (AlluxioFormat) Class.forName(formatClassName).newInstance();
+    } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
+      LOG.error("Exception is {}", e);
+      throw new RuntimeException("Failed to initiate class" + formatClassName);
```
We can use `ConnectException` here so that Kafka Connect knows how to properly catch it.

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,177 @@
+      try {
+        writer.writeRecord();
+      } catch (IOException | AlluxioException e) {
+        LOG.error("Failed to write record to " + writer.getTopicPartition().toString());
```
In case of exception, do we just throw the SinkRecords? This can potentially lose data. We need a mechanism to handle transient exceptions and fail the connector in case of the

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,177 @@
+   * @param offsets mapping of TopicPartition to committed offset
+   */
+  @Override
+  public void flush(Map<TopicPartition, OffsetAndMetadata> offsets) {
```
Kafka Connect calls flush before committing offsets in Kafka for this connector, in order to achieve at least once delivery we need to ensure that the data is already persisted in Alluxio before committing offset in Kafka.  

The HDFS connector is different as it manages the offsets on its own.

yupeng9 replied:
@Ishiihara thanks for pointing it out. Can you be more specific on how to implement this `flush` method? Is there any additional logic to be added?

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,177 @@
+   */
+  @Override
+  public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+    //TODO(GuangHui): Create writers for newly assigned partitions in case of partition rebalance
```
These are really important to achieve any delivery guarantee. It will be really nice to add them before merging to master. @zhutongxue Let me know if you want to discuss more.

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkTask.java:
```diff
@@ -0,0 +1,177 @@
+   * @return Connector version
+   */
+  @Override
+  public String version() {
```
Same comment as in AlluxioSinkConnector.

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioTopicPartitionWriter.java:
```diff
@@ -0,0 +1,234 @@
+  private long mLastRotationTime;
+  private long mRotationRecordNum;
+  private long mRotationTimeInterval;
+  private Queue<SinkRecord> mRecordQueue = new LinkedBlockingQueue<>();
```
Is the writer used by multiple threads? Seems only used by one task. Probably used a LinkedList instead?

Ishiihara commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioTopicPartitionWriter.java:
```diff
@@ -0,0 +1,234 @@
+   * @param record SinkRecord
+   */
+  public void writeToBuffer(SinkRecord record) {
+    mRecordQueue.offer(record);
```
We need some memory management here, otherwise, the JVM can OOM.

yupeng9 replied:
@Ishiihara can you give a pointer to how other connectors handle this issue?

Ishiihara commented:
@zhutongxue @apc999 Thank for contributing the Alluxio connector with Kafka Connect. I left a couple of comments.  
Some of the comments can be addressed in following PRs. But I think it will be nice if we can get the the at lease once delivery guarantee in the connector.  To get there, we need to properly handle some exceptions and correct implementation of flush method in the sink task.  Happy to discuss more, @zhutongxue If you want, you can reach to @apc999 to get my contact info. We can have a call to discuss further.

zhutongxue commented:
@Ishiihara Thank you very much for your comments, the comments are very important and worth careful consideration.

I am very glad to discuss with you and @apc999 about the details in order to achieve at least once delivery. As you said, some exceptions should be handled properly. Happy to discuss more.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10142/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10143/
Test PASSed.

zhutongxue commented:
@Ishiihara  @apc999   I  have pushed some commits about Kafka Alluxio connect. 

To achieve at least once delivery guarantee,  the exception of writing record to Alluxio are handled. Furthermore, open() and close() method are implemented in  AlluxioSinkTask.
The new commits also integrate  Confluent Platform 3.0 with Alluxio connect. 
Readme file  has also been modified as you suggest.

Happy to discuss more. Thank you !

apc999 commented:
@zhutongxue thanks for the update. I see more unit tests here. Will have a more careful review after 1.1 released.

apc999 commented:
@zhutongxue I think I accidentally closed this PR. Sorry for the mistake. It seems the branch has been deleted?

apc999 commented:
Actually, I don't even remember I closed this PR. I don't have a clue here what's happening here and why this branch has been deleted. Could you reopen the branch and let's resume this PR? Sorry about the mess.

zhutongxue commented:
@apc999  That's nothing. Of Course, let's continue working on this job and complete the PR.

zhutongxue commented:
@apc999  The connect branch has been deleted , So I don't know where to reopen the branch. Maybe we should first create the connect branch.

apc999 commented:
@zhutongxue I see why now. when I was cleaning the repository for release 1.1, I accidentally removed this branch in upstream with other unused branches. It is my bad. sorry for that. Now I am re-created this branch and let's re-open this PR.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/10345/
Test PASSed.

apc999 commented:
@Ishiihara would you like to take a pass on this update?

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnectorConfig.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Map;
+
+/**
+ * This class is responsible for reading and storing allxuio connector config parameters.
```
nit: allxio -> Alluxio

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnectorConfig.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * This class is responsible for reading and storing allxuio connector config parameters.
+ */
+public class AlluxioSinkConnectorConfig extends AbstractConfig {
```
if no other class extending this one, Alluxio style marks it `final`

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnectorConfig.java:
```diff
@@ -0,0 +1,74 @@
+          RETRY_TIME_INTERVAL_MS_DOC);
+
+  /**
+   * Gets ConfigDef.
```
NIT: for trivial getters, you can simply have `@return` as javadoc

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/format/JsonFormat.java:
```diff
@@ -0,0 +1,61 @@
+ */
+public class JsonFormat implements AlluxioFormat {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JsonFormat.class);
```
Alluxio style recommends to do ` private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);`

apc999 commented on integration/kafka/src/main/java/alluxio/kafka/connect/format/JsonFormat.java:
```diff
@@ -0,0 +1,61 @@
+  public void writeRecord(FileOutStream fileStream, SinkRecord record)
+      throws IOException {
+    if (record.value() != null) {
+      LOG.info(record.value().toString());
```
those `LOG`s are on echo write operation. shall we avoid logging too much?

apc999 replied:
e.g., you may change level from `info` to `debug`

apc999 commented on integration/kafka/src/test/java/alluxio/kafka/connect/AlluxioSinkConnectorTest.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Tests for the {@link AlluxioSinkConnector} class..
+ */
+public class AlluxioSinkConnectorTest {
```
`final` please

yupeng9 commented on build/findbugs/findbugs-exclude.xml:
```diff
@@ -12,7 +12,12 @@
+  
+  <Match>
+    <Package name="alluxio.kafka.connect"/>
+    <Bug code="SF"/>
```
Why do you need to disable FB for this bug code?

zhutongxue replied:
In writeRecord method of AlluxioTopicPartitionWriter class , the switch-case statement is used to write data to Alluxio file. The implementation in each case corresponds to the current state. When the current state has been processed, the next state begins to be processed. So ,I didn't add beak statement in the end of case processing. But with this method , the build can not be passed . So , I add SF code for kafka-connect in findbugs-exculde.xml.  
Sorry , I didn't leave comments here before.

yupeng9 replied:
Sounds reasonable. Thanks.

yupeng9 commented on integration/kafka/ReadMe.md:
```diff
@@ -0,0 +1,110 @@
+##Kafka Alluxio Connector
+Kafka Alluxio Connector exports data in kafka to Alluxio filesystem. The connector supports user defined data format and shipped with JSON Format.
+##Perquisites
+- [confluent-3.0](http://www.confluent.io/download)
```
Does this connector work with vanilla Kafka?

zhutongxue replied:
yes, Kafka latest 0.10 version required.

zhutongxue replied:
I will try to change Perquisites  confluent-3.0 to Kafka.

yupeng9 replied:
sounds good

yupeng9 commented on integration/kafka/pom.xml:
```diff
@@ -0,0 +1,88 @@
+  <parent>
+    <groupId>org.alluxio</groupId>
+    <artifactId>alluxio-integration</artifactId>
+    <version>1.1.0-SNAPSHOT</version>
```
Can you update this to `1.2.0-SNAPSHOT`

zhutongxue replied:
of course ok

zhutongxue replied:
But, the connect branch is still 1.1.0-SNAPSHOT. So , if I push to connect branch, someting maybe confilct.

yupeng9 replied:
You can pull master into your branch. Given your change is quite isolated, there shouldn't be much conflict.

yupeng9 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnectorConfig.java:
```diff
@@ -0,0 +1,74 @@
+  public static final String ROTATION_TIME_INTERVAL_MS = "rotation.time.interval.ms";
+  public static final String ALLUXIO_FORMAT = "alluxio.format";
+  public static final int OFFSET_LENGTH = 10;
+  public static final String ALLUXIO_CONNECTOR_VERSION = "1.1.0";
```
You can get Alluxio version from `RuntimeConstants.VERSION`

yupeng9 commented on integration/kafka/src/main/java/alluxio/kafka/connect/AlluxioSinkConnectorConfig.java:
```diff
@@ -0,0 +1,74 @@
+  private static final String RETRY_TIME_INTERVAL_MS_DOC =
+      "The retry interval in case of transient exception";
+  private static ConfigDef sConfig = new ConfigDef()
+      .define(ALLUXIO_URL, Type.STRING, "alluxio://localhost:19998", ConfigDef.Importance.HIGH,
```
Instead of hardcoding the hostname and port, it's better to read them from the configurations.

zhutongxue replied:
yes, when the connector runs, it will read configuration file. The hostname and port in the above code is default value when the configuration file can not be found.

yupeng9 commented:
Thanks @zhutongxue for adding this awesome connector. I left some comments. Also, there're some comments from @apc999 and @Ishiihara unaddressed. Please take a look.

yupeng9 commented:
Hi, @zhutongxue any updates on this?

zhutongxue commented:
@yupeng9  Sorry for reply so later. I will address these comments as soon as possible.

zhutongxue commented:
@apc999  Thanks for your comments. Sorry to reply so later. Should I continue push the connector code to connect branch?

enchantner commented:
Hey guys, what's the status of this? We are looking forward to using Kafka with Alluxio, but all I found is this old pull request.

zhutongxue commented:
@enchantner  Hey, sorry to rely later.  Yes, We are still working on integrating Kafka with Alluxio and we have implemented a preliminary version.  But I don't know the current implementation is able to satisfy your requirements. So, Would you describe your use case first, and  then we could discuss the details together.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/13909/
Test PASSed.
